### PR TITLE
KAFKA-10777: Add additional configuration to control MirrorMaker 2 internal topics naming convention

### PR DESCRIPTION
**Pull request by OmniaGM**
[KIP-690](https://cwiki.apache.org/confluence/display/KAFKA/KIP-690%3A+Add+additional+configuration+to+control+MirrorMaker+2+internal+topics+naming+convention)  

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by OmniaGM** [comment]
@ryannedolan & @mimaison can I get a review on this, please?

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/DefaultReplicationPolicy.java [hunk 1]
```diff
@@ -70,4 +70,32 @@ public String upstreamTopic(String topic) {
             return topic.substring(source.length() + separator.length());
         }
     }
+
+    private String internalSuffix() {
+        return separator + "internal";
+    }
+
+    private String checkpointTopicSufficx() {
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/DefaultReplicationPolicy.java]: `checkpointTopicSufficx` -> `checkpointsTopicSuffix`

**Comment by mimaison** [comment]
Thanks @OmniaGM for the PR. I'll try to take a look next week.

**Comment by OmniaGM** [comment]
Hi @mimaison,  did you get a chance to have a look into this pr?

**Comment by mimaison** [comment]
@OmniaGM Sorry for the delay, I was away for a few weeks. I'm back now and I'll take a look

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java [hunk 2]
```diff
@@ -52,9 +52,39 @@ default String originalTopic(String topic) {
...
+    }
+
+    /** Returns the offset-syncs topic for given cluster alias. */
+    default String offsetSyncTopic(String clusterAlias) {
+        return "mm2-offset-syncs." + clusterAlias + ".internal";
+    }
+
+    /** Returns the name checkpoint topic for given cluster alias. */
+    default String checkpointTopic(String clusterAlias) {
+        return clusterAlias + ".checkpoint.internal";
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: This topic is currently named `.checkpoints.internal`, see https://github.com/apache/kafka/blob/trunk/connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClientConfig.java#L60. We need to keep the current name

Comment (mimaison) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: Can we also rationalize the different names. Ig uess this method should be `checkpointsTopic`.

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java [hunk 3]
```diff
@@ -52,9 +52,39 @@ default String originalTopic(String topic) {
...
+    }
+
+    /** check if topic is a heartbeat topic, e.g heartbeats, us-west.heartbeats. */
+    default boolean isHeartbeatTopic(String topic) {
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: Should we call this method `isHeartbeatsTopic` instead, so it lines up with `heartbeatsTopic`?

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java [hunk 4]
```diff
@@ -52,9 +52,39 @@ default String originalTopic(String topic) {
         }
     }
 
+    /** Returns heartbeats topic name.*/
+    default String heartbeatsTopic() {
+        return "heartbeats";
+    }
+
+    /** Returns the offset-syncs topic for given cluster alias. */
+    default String offsetSyncTopic(String clusterAlias) {
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: Should we called this method `offsetSyncsTopic` so it matches the actual topic name?

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/DefaultReplicationPolicy.java [hunk 5]
```diff
@@ -70,4 +70,32 @@ public String upstreamTopic(String topic) {
...
+    private String internalSuffix() {
+        return separator + "internal";
+    }
+
+    private String checkpointsTopicSuffix() {
+        return separator + "checkpoints" + internalSuffix();
+    }
+
+    @Override
+    public String offsetSyncTopic(String targetAlias) {
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/DefaultReplicationPolicy.java]: Should we name the argument `clusterAlias` as it's not necessarily the target cluster alias?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a few comments. I suggested alternative names for a few methods, as these were part of the KIP, if you agree, we'll need to update it.

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java (hunk 2 above)
Comment (OmniaGM) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: > This topic is currently named .checkpoints.internal, see https://github.com/apache/kafka/blob/trunk/connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClientConfig.java#L60. We need to keep the current name

Good spot

> Can we also rationalize the different names. Ig uess this method should be checkpointsTopic.

make sense, will update the KIP as well to reflect this.

Comment (mimaison) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: Can you also send an update to the VOTE thread? so all participants are notified. Thanks

Comment (OmniaGM) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/ReplicationPolicy.java]: Thanks for sending the update to the VOTE thread 🙇🏼

**Comment by mimaison** [comment]
@OmniaGM I sent an update to the VOTE thread of your KIP: https://lists.apache.org/thread.html/rbdcdfe49229f7e1d2684a1a5f17364004f8b7545cd65bc35d296e327%40%3Cdev.kafka.apache.org%3E

If nobody complains, I'll merge the PR early next week.
Thanks again

**Review by mimaison — APPROVED** [review]
LGTM

**Comment by mimaison** [comment]
Thanks @OmniaGM for this contribution! Sorry it took so long but we got there in the end 🙂  Looking forward to seeing your next contributions

**Comment by OmniaGM** [comment]
> Hello, I know this PR is merged but I have an issue related to it while implementing a custom ReplicationPolicy to override the `heartbeat` topic name. According to the resources below, In order to achieve it, It's necessary to create a jar file and to reference it in the `plugin.path` properties file. Unfortunately it didn't work. What worked was setting the jar file path in the `CLASSPATH` environment variable. So either such config is missing in the documentation or I am doing something wrong in which case I would be thankful if you point it out. Resources : https://cwiki.apache.org/confluence/display/KAFKA/KIP-690%3A+Add+additional+configuration+to+control+MirrorMaker+2+internal+topics+naming+convention https://stackoverflow.com/questions/59390555/is-it-possible-to-replicate-kafka-topics-without-alias-prefix-with-mirrormaker2 I will be more than happy to provide more context.

Hi @ahibot , 
A similar issue to what you are facing was reported back in 2020 against Kafka 2.4 (details [here](https://issues.apache.org/jira/browse/KAFKA-9376) with Connect not able to load MM2 `plugin.path` and @ryannedolan mentioned that he couldn't reproduce the same issue at the time.  So I don't believe this issue you are describing is related to this PR and KIP-690 which was released in Kafa [3.1](https://blogs.apache.org/kafka/entry/what-s-new-in-apache7). 

However, I can see that the comments on [KAFKA-9376](https://issues.apache.org/jira/browse/KAFKA-9376) suggest `CLASSPATH` as a workaround and there is even a comment mentioning that they can't see this issue with 2.5 anymore. So if you are running version >= 2.5 then I would recommend commenting on [KAFKA-9376](https://issues.apache.org/jira/browse/KAFKA-9376) with more details regarding how to reproduce this issue and which version you are using.

Edit: You may need to check that all your MM2 connectors and the dependencies are under `plugin.path`. 

Thanks

**Pull request merged**